File: 프로그래머스/3/67259. ［카카오 인턴］ 경주로 건설/［카카오 인턴］ 경주로 건설.cpp
#include <string>
#include <vector>
#include <queue>
#include <climits>
#include <algorithm>

using namespace std;

int dx[4] = { 1, 0, -1, 0 };
int dy[4] = { 0, 1, 0, -1 };

bool is_corner(int curDir, int nxtDir) {
    return (curDir % 2) != (nxtDir % 2);
}

int solution(vector<vector<int>> board) {
    int answer = 0;
    int N = board.size();
    
    queue<pair<int, pair<int, int>>> q; // {온 방향, {x, y}}
    vector<vector<vector<int>>> cost(N, vector<vector<int>>(N, vector<int>(4, INT_MAX)));
    
    cost[0][0][0] = cost[0][0][1] = cost[0][0][2] = cost[0][0][3] = 0;
    if(board[1][0] == 0) {
        q.push({0, {1, 0}});
        cost[1][0][0] = 100;
    }
    if(board[0][1] == 0) {
        q.push({1, {0, 1}});
        cost[0][1][1] = 100;
    }
    
    while(!q.empty()) {
        auto curInfo = q.front();
        q.pop();
        
        int cur_dir = curInfo.first;
        int curX = curInfo.second.first;
        int curY = curInfo.second.second;
        
        for(int dir = 0; dir < 4; dir++) {
            int nxtX = curX + dx[dir];
            int nxtY = curY + dy[dir];
            
            if(nxtX < 0 || nxtX >= N || nxtY < 0 || nxtY >= N) continue;
            if(board[nxtX][nxtY] == 1) continue;
            
            int weight;
            if(is_corner(cur_dir, dir)) weight = 600;
            else weight = 100;
            
            if(cost[nxtX][nxtY][dir] > cost[curX][curY][cur_dir] + weight) {
                cost[nxtX][nxtY][dir] = cost[curX][curY][cur_dir] + weight;
                q.push({dir, {nxtX, nxtY}});
            }
        }
    }
    
    return min(cost[N - 1][N - 1][0], cost[N - 1][N - 1][1]);
}

File: 프로그래머스/3/67259. ［카카오 인턴］ 경주로 건설/［카카오 인턴］ 경주로 건설.py
import sys
from collections import deque

INF = sys.maxsize

dx = [1, 0, -1, 0] # 아래, 오른쪽, 위, 왼쪽
dy = [0, 1, 0, -1]

def is_out(x, y, n):
    if x < 0 or x >= n or y < 0 or y >= n:
        return True
    return False

def is_vertical(l_dir, r_dir):
    if l_dir % 2 != r_dir % 2:
        return True
    return False

def solution(board):
    n = len(board)
    cost = [[[INF] * 4 for _ in range(n)] for _ in range(n)]
    q = deque()
    
    for idx in range(4):
        cost[0][0][idx] = 0
        
    if board[0][1] != 1:
        cost[0][1][1] = 100
        q.append((0, 1, 1))
    if board[1][0] != 1:
        cost[1][0][0] = 100
        q.append((1, 0, 0))
    
    while q:
        curX, curY, prevDir = q.popleft()
        
        for curDir in range(4):
            nxtX = curX + dx[curDir]
            nxtY = curY + dy[curDir]
            
            if is_out(nxtX, nxtY, n) or board[nxtX][nxtY] == 1:
                continue
            
            if is_vertical(prevDir, curDir):
                nxtCost = cost[curX][curY][prevDir] + 600
            else:
                nxtCost = cost[curX][curY][prevDir] + 100

            if nxtCost < cost[nxtX][nxtY][curDir]:
                cost[nxtX][nxtY][curDir] = nxtCost
                q.append((nxtX, nxtY, curDir))
    
    return min(cost[n - 1][n - 1][0], cost[n - 1][n - 1][1])